v13: add curation checks documentation (#796)
inveniosoftware/docs-invenio-rdm: docs/reference/checks_config.md

# Metadata Checks Configuration Schema

This documentation explains the schema and configuration format for creating metadata validation rules in the checks system.

## Overview

> The system provides a flexible way to define validation rules for record metadata using JSON stored in the database. Rules can check individual fields, compare values, and validate lists of items with complex conditions. This schema provides a powerful way to define complex validation logic while remaining human-readable and maintainable.

Reviewer: ```suggestion
The system provides a flexible way to define validation rules for record metadata using JSON stored in the database. Rules can check individual fields, compare values, and validate lists of items with complex conditions. This schema provides a powerful way to define complex validation logic while staying human-readable and maintainable.
```

## Rule Configuration

A rule configuration is a dictionary with the following structure:

```python
{
    "id": "unique-rule-id",  # Required
    "title": "Human-readable title",
    "message": "Error message shown when rule fails",
    "description": "Detailed description of the rule",
    "level": "info|warning|failure",  # Default: "info"
    "condition": {  # Optional condition expression
        "type": "expression_type",
        # ... expression configuration
    },
    "checks": [  # List of check expressions
        {
            "type": "expression_type",
            # ... expression configuration
        }
    ]
}
```

> This object is stored in the `params` section of the check_config in the database.

Reviewer: ```suggestion
This object is stored in the `params` section of the `CheckConfig` model in the database.
```

> ## Expression Types

Reviewer: I feel there's a small "prologue" section needed here to give some context about what an "expression" is, e.g. comparing it to Excel cell formulas or a scripting language "eval".

### 1. Field Expression

> Accesses a fiel, or nested field, from the metadata using dot notation, returning that value to the check.

Reviewer: ```suggestion
Accesses a field from the metadata using dot notation, returning that value.
```

```python
{
    "type": "field",
    "path": "metadata.title"  # Path to the field
}

> ```

Reviewer: I'm not sure how I feel about defining these "schema"-like structures as code-blocks vs. tables. An alternative would be to do something similar to JSDoc-like code blocks, see e.g. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#command

Author: also I think representing this as a table would be more confusing as you write and edit these in JSON

                        "operator": "and",
                        "expressions": [
                            {
                                "path": "person_or_org.identifiers",
                                "type": "field"
                            },
                            {
                                "path": "person_or_org.identifiers",
                                "type": "list",
                                "operator": "any",
                                "predicate": {
                                    "path": "identifier",
                                    "type": "field"
                                }
                            }
                        ]
                    }
                }
            ],
            "message": "Affiliations are recommended for all creators (e.g. an ORCID)",
            "condition": {
                "path": "metadata.creators",
                "type": "list",
                "operator": "exists",
                "predicate": {
                }
            },
            "description": "All creators should have a persistent identifier (e.g. an ORCID)"
        },
    ]
}

> ```

Reviewer: Here, I would also expect an example of what kind of result comes back from evaluating the above rules against a record.

> # Metadata Checks Configuration Schema

Reviewer 2: *placement discussion* : would maybe place this in "Maintain and Develop" (even if it is a reference). It's not a general reference usable by many different kinds of users right? Maintainers and maybe operators only. Keeping it close to the other relevant doc makes it easier to grasp the concept in a one-stop-shop location.
